### PR TITLE
Add filter to griddata

### PR DESCRIPTION
**Pull request by JacquesBorsenberger**

### CODE REVIEW

File: src/utils/rest-api.js [hunk 1]
```diff
@@ -246,3 +248,64 @@ export function addFiltersContingencyList(name, newFiltersContingency) {
...
+export function saveFilter(newFilter) {
+    return backendFetch(PREFIX_FILTERS_QUERIES, {
+        method: 'put',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(newFilter),
+    });
+}
+
+/**
+ * Get all contingency lists
```
Comment (FranckLecuyer) [inline src/utils/rest-api.js]: Change comment

File: src/utils/rest-api.js [hunk 2]
```diff
@@ -246,3 +248,64 @@ export function addFiltersContingencyList(name, newFiltersContingency) {
...
+ * @returns {Promise<Response>}
+ */
+export function getFilters() {
+    return backendFetch(PREFIX_FILTERS_QUERIES)
+        .then((response) => response.json())
+        .then((res) => res.sort((a, b) => a.name.localeCompare(b.name)));
+}
+
+/**
+ * Delete contingency list by name
```
Comment (FranckLecuyer) [inline src/utils/rest-api.js]: Change comment

File: src/utils/rest-api.js [hunk 3]
```diff
@@ -246,3 +248,64 @@ export function addFiltersContingencyList(name, newFiltersContingency) {
...
+ */
+export function deleteFilterByName(name) {
+    const url = PREFIX_FILTERS_QUERIES + encodeURIComponent(name);
+    return backendFetch(url, {
+        method: 'delete',
+    });
+}
+
+/**
+ * Get contingency list by type
```
Comment (FranckLecuyer) [inline src/utils/rest-api.js]: Change comment

File: src/utils/rest-api.js [hunk 4]
```diff
@@ -246,3 +248,64 @@ export function addFiltersContingencyList(name, newFiltersContingency) {
...
+ * @returns {Promise<Response>}
+ */
+export function getFilterByName(name) {
+    const url = PREFIX_FILTERS_QUERIES + name;
+    return backendFetch(url).then((response) => response.json());
+}
+
+/**
+ * Rename list by name
```
Comment (FranckLecuyer) [inline src/utils/rest-api.js]: Change comment

File: src/translations/en.json [hunk 5]
```diff
@@ -22,11 +28,14 @@
...
     "nominalVoltage": "Nominal Voltage",
+    "Generators": "Generators",
+    "Lines": "Lines",
+    "Min": "Min",
+    "Max": "Max",
     "SCRIPT": "Script",
     "FILTERS": "Filters",
     "TWO_WINDINGS_TRANSFORMER": "2W TRANSFORMER",
     "DANGLING_LINE": "DANGLING LINE",
-    "LINE": "LINE",
```
Comment (FranckLecuyer) [inline src/translations/en.json:31]: "LINE" should remain

File: src/translations/fr.json [hunk 6]
```diff
@@ -4,6 +4,11 @@
     "addNewContencyFile": "Nom de la nouvelle liste",
```
Comment (FranckLecuyer) [inline src/translations/fr.json]: Coherence problems between french and english json files
Ex : "Min", "Max"  <---> "MinNominalVoltage", "MaxNominalVoltage"

File: src/components/data-tabs.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                <ContingencyLists
+                    FilterRenderer={FiltersEditor}
+                    listType={'contingencies'}
+                />
+            ),
+        },
+        {
+            name: 'FILTERS',
+            render: () => (
+                <FilterList FilterRenderer={GenericFilter} listType={'line'} />
```
Comment (FranckLecuyer) [inline src/components/data-tabs.js]: unused props

File: src/components/contingency-list.js [hunk 8]
```diff
@@ -196,7 +185,7 @@ const CustomTooltip = (props) => {
     return <Tooltip arrow classes={classes} {...props} />;
 };
 
-const ContingencyLists = () => {
+const ContingencyLists = ({ listType, FilterRenderer }) => {
```
Comment (FranckLecuyer) [inline src/components/contingency-list.js]: Don't know the utility of these props

File: src/components/contingency-list.js [hunk 9]
```diff
@@ -424,7 +413,7 @@ const ContingencyLists = () => {
         replaceFiltersWithScriptContingencyList(currentItem.name)
             .then((response) => {
                 if (response.ok) {
-                    getContingencyLists().then((data) => {
+                    getContingencyLists(listType).then((data) => {
```
Comment (FranckLecuyer) [inline src/components/contingency-list.js]: getContingencyLists has no argument

File: src/components/filter-list.js [hunk 10]
```diff
@@ -0,0 +1,427 @@
+/**
+ * Copyright (c) 2020, RTE (http://www.rte-france.com)
```
Comment (FranckLecuyer) [inline src/components/filter-list.js]: 2021

File: src/components/data-tabs.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import React, { useState } from 'react';
```
Comment (FranckLecuyer) [inline src/components/data-tabs.js:8]: Missing copyright

File: src/components/generic-filter.js [hunk 12]
```diff
@@ -0,0 +1,246 @@
+import { FormattedMessage } from 'react-intl';
```
Comment (FranckLecuyer) [inline src/components/generic-filter.js:8]: Missing copyright

File: src/components/filters.js [hunk 13]
```diff
@@ -0,0 +1,163 @@
+import React, { useRef, useState } from 'react';
```
Comment (FranckLecuyer) [inline src/components/filters.js]: Missing copyright

File: src/components/custom-list-item.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
+import { ScriptTypes } from '../utils/script-types';
```
Comment (FranckLecuyer) [inline src/components/custom-list-item.js:8]: Missing copyright

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

**Comment by FranckLecuyer** [comment]
When we change nominal voltage filter from 'range' or 'approx' to 'equality', the 'value2' field remains unchanged in the sql table 'numericfilter', while it should be deleted

**Comment by FranckLecuyer** [comment]
Labels 'equality', 'range', 'approx' are not aligned with the combobox arrow

File: src/translations/fr.json [hunk 15]
```diff
@@ -39,5 +46,16 @@
     "replaceList": "Remplacer la liste",
     "alertBeforeReplaceWithScript": "Voulez-vous vraiment remplacer cette liste par un script Groovy ?{br}{br}Cette action est irréversible",
     "remplacer": "Remplacer",
-    "copy": "Copier"
+    "copy": "Copier",
+    "newFilter": "Nouveau filtre",
+    "FilterName": "Nom",
+    "addNewFilter": "Créer un filtre",
+    "substationName": "Nom Poste",
```
Comment (FranckLecuyer) [inline src/translations/fr.json]: If it is 'substationName, french translation should be 'Nom Site'
If it is 'Nom Poste', the english translation should be 'VoltageLevel Name' and the key renamed

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

**Review by FranckLecuyer — APPROVED** [review]

File: src/components/filter-list.js [hunk 16]
```diff
@@ -0,0 +1,427 @@
...
+                existingList={filterList}
+                action={({ name }) => renameList(currentItemName, name)}
+                {...props}
+            />
+        ),
+        deleteFilter: ({ ...props }) => (
+            <PopupInfo
+                title={<FormattedMessage id="deleteFilter" />}
+                customAlertMessage={
+                    <FormattedMessage id="alertBeforeDeleteFilter" />
```
Comment (niconoir) [inline src/components/filter-list.js:312]: Message translation alertBeforeDeleteFilter is not present in translations files (fr.json and en.json)

File: src/translations/en.json [hunk 17]
```diff
@@ -39,5 +49,12 @@
...
     "remplacer": "Replace",
-    "copy": "Copy"
+    "copy": "Copy",
+    "newFilter": "New filter",
+    "FilterName": "Filter Name",
+    "addNewFilter": "Add a new filter",
+    "substationName": "Substation Name",
+    "renameFilter": "Rename filter",
+    "newFilterName": "new filter name",
+    "deleteFilter": "Delete filter"
```
Comment (niconoir) [inline src/translations/en.json:62]: Missing "alertBeforeDeleteFilter" translation

File: src/translations/fr.json [hunk 18]
```diff
@@ -39,5 +46,16 @@
...
+    "newFilter": "Nouveau filtre",
+    "FilterName": "Nom",
+    "addNewFilter": "Créer un filtre",
+    "substationName": "Nom Site",
+    "renameFilter": "Renommer le filtre",
+    "newFilterName": "nouveau nom",
+    "deleteFilter": "supprimer le filtre",
+    "equality": "egalité",
+    "approx": "approximation",
+    "range": "interval"
```
Comment (niconoir) [inline src/translations/fr.json]: Missing "alertBeforeDeleteFilter" translation

File: src/redux/actions.js [hunk 19]
```diff
@@ -36,3 +36,13 @@ export function updateContingencyList(contingencyLists) {
         contingencyLists: contingencyLists,
     };
 }
+
+export const UPDATE_FILTER_LIST = 'UPDATE_FILTER_LIST';
+
+export function updateFilterList(filterList) {
+    console.info(filterList);
```
Comment (niconoir) [inline src/redux/actions.js]: Is this console logging necessary ?

File: src/components/filter-list.js [hunk 20]
```diff
@@ -0,0 +1,427 @@
...
+        if (originalFilter !== null) setCurrentEdit(originalFilter);
+    };
+
+    const newFilter = (name, type) => {
+        let newFilterList = [...filterList];
+        setCurrentItemName(name);
+        const newFilter = {
+            name: name,
+            type: type === ScriptTypes.SCRIPT ? type : 'LINE',
+            transient: true,
```
Comment (niconoir) [inline src/components/filter-list.js:208]: Why create the filter as transient ? It makes the creation behaviour different from contingency creation.

File: src/components/filter-list.js [hunk 21]
```diff
@@ -0,0 +1,427 @@
...
+                {...props}
+            />
+        ),
+    };
+
+    function renderList() {
+        return (
+            <>
+                <Button
+                    className={classes.addNewList}
```
Comment (niconoir) [inline src/components/filter-list.js:326]: Here, the button style is quite different from the style used for new list creation in contingency-list. They should be identical, for the contingency list and filter list to look identical (right now, the buttons icons and labels sizes are different).

File: src/components/filters.js [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+    );
+};
+
+export const CountriesSelection = ({ initialValue, onChange, disabled }) => {
+    let countriesList;
+    try {
+        countriesList = require('localized-countries')(
+            require('localized-countries/data/' +
+                navigator.language.substr(0, 2))
+        );
```
Comment (niconoir) [inline src/components/filters.js]: The language which is used for countries names localization should be retrieved in the application parameters, it is not necessarily the language of the navigator.

File: src/components/generic-filter.js [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+                    );
+                else return renderFilter(key, definition);
+            }
+        );
+    };
+
+    return (
+        <Grid
+            container
+            direction="row"
```
Comment (niconoir) [inline src/components/generic-filter.js]: Why "row" and not "column" ?

File: src/components/generic-filter.js [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+                        <Grid
+                            container
+                            item
+                            direction={'row'}
+                            key={key}
+                            spacing={1}
+                        >
+                            {[
+                                ...Array.from(Array(definition.occurs).keys()),
+                            ].map((n) => {
```
Comment (niconoir) [inline src/components/generic-filter.js:211]: The problem with this way to generate filters components is that it creates multiple filters with exactly the same label, and thus, we don't exactly know what the filter is used for (i.e. we can have two "countries" filters, and we don't exactly know what they really do)

**Review by niconoir — CHANGES_REQUESTED** [review]
Contingency list and filter list are implemented in quite different ways, but they look and behave quite the same way. Maybe it would be useful to refactor contingency-list component to make its implementation closer to the filter-list component implementation. This would avoid some bugs like the look of the "New filter" and "New list" button which should be the same but actually look different.

File: src/components/filter-list.js (hunk 21 above)
Comment (JacquesBorsenberger) [reply src/components/filter-list.js:326]: need a whole US to refactor the contingency list

File: src/components/filters.js (hunk 22 above)
Comment (JacquesBorsenberger) [reply src/components/filters.js]: done

File: src/components/generic-filter.js (hunk 23 above)
Comment (JacquesBorsenberger) [reply src/components/generic-filter.js]: removed

**Review by niconoir — APPROVED** [review]

**Pull request merged**